data/calculated_parser/functions.py

```diff
@@ -69,11 +69,313 @@ def sspeed(depth, latitude, temperature, salinity):
+    cd_depth = depth.values[cd_idx]
+
+    if cd_value < sld_value:
+        
+        # Find 2 Points
+        cd_value_1 = cd_value
+        cd_value_2 = speed_point[cd_idx + 1]
+
+        cd_depth_1 = depth.values[cd_idx]
+        cd_depth_2 = depth.values[cd_idx + 1]
+
+        cd_depth = linearInterp(cd_value_1, cd_depth_1, cd_value_2, cd_depth_2, sld_value)
+
+    elif cd_value > sld_value:
+        
+        # Find 2 Points
+        cd_value_1 = speed_point[cd_idx - 1]
+        cd_value_2 = cd_value
+
+        cd_depth_1 = depth.values[cd_idx - 1]
+        cd_depth_2 = depth.values[cd_idx]
+
+        cd_depth = linearInterp(cd_value_1, cd_depth_1, cd_value_2, cd_depth_2, sld_value)
+
+    return cd_depth
+
+def soundchannelaxis(depth, lat, temperature, salinity):
+    """
+    Finds the global minimum of the speed of sound
+
+    Parameters:
+    depth: The depth(s) in meters
+    lat: The latitude(s) in degrees North
+    temperature: The temperatures(s) (at all depths) in celsius
+    salinity: The salinity (at all depths) (unitless)
+    """
+    
+    speed = sspeed(depth, lat, temperature, salinity)
+    for x in range(speed.shape[-1]):
+        for y in range(speed.shape[-2]):
+            if (speed[:,y,x].size - np.count_nonzero(np.isnan(speed[:,y,x]))) != 0:
+                idx = find_sca_idx(speed[:,y,x])
+                if np.isnan(idx):
+                    speed[:,y,x] = idx
+                else:
+                    speed[:,y,x] = depth.values[idx]
+            else:
+                speed[:,y,x] = np.nan
+                
+    return speed[0]
+
+def soniclayerdepth(depth, lat, temperature, salinity):
+    """
+    Finds the local maxima of the speed of sound (local maxima is before the sound channel axis)
+
+    Parameters:
+    depth: The depth(s) in meters
+    lat: The latitude(s) in degrees North
+    temperature: The temperatures(s) (at all depths) in celsius
+    salinity: The salinity (at all depths) (unitless)
+
+    Note:
+    sld is the critical depth (the DEPTH at which the sspeed local maxima occurs)
+    sld_value, is the value of the sound speed at the critical depth
+    """
+    
+    # Find speed of sound
+    speed = sspeed(depth, lat, temperature, salinity)
+    
+    for x in range(speed.shape[-1]):
+        for y in range(speed.shape[-2]):
+            if (speed[:,y,x].size - np.count_nonzero(np.isnan(speed[:,y,x]))) != 0:
+            
+                sca_idx = find_sca_idx(speed[:,y,x])
+
+                if (np.isnan(sca_idx)):
+                    speed[:,y,x] = sca_idx
+                else:
+                    sld_idx = find_sld_idx(sca_idx, speed[:,y,x])
+
+                    if (np.isnan(sld_idx)):
+                        speed[:,y,x] = sld_idx
+                    else:
+                        sld = depth.values[sld_idx]
+                        speed[:, y, x] = sld
+            else:
+                speed[:,y,x] = np.nan
+
+    return speed[0] # Only return one horizontal slice
```

Comment:
From the meeting I would only suggest altering the first bullet I discussed.

Reply:
I agree with Nabil on the first bullet. I was confused by modifying the speed array on my first pass through the code. It would be more readable to create a new array and add to that. 

On the second bullet, I think it will take some effort to vectorize and that should come in a second revision. When we get there, we might consider functions like numpy.nanargmin. My concern is that this will through a ValueError if it encounters a slice with all NaNs which could happen over land.
https://docs.scipy.org/doc/numpy/reference/generated/numpy.nanargmin.html

Reply:
Partially done (the part that was requested)

Reply:
Created a new issue for vectorization